### ISSUE TITLE
Reproducibility of MOOP by setting the seed of random number generators

### ISSUE THREAD

**Issue by hija** [jMetal/jMetal#155]
Hello everyone,

I have a question concerning multi objective optimization using the jMetal Framework. I have used the "[NSGAIITSPRunner.java](https://github.com/jMetal/jMetal/blob/master/jmetal-exec/src/main/java/org/uma/jmetal/runner/multiobjective/NSGAIITSPRunner.java)" (from the package org.uma.jmetal.runner.multiobjective and jMetal 5.1) and for adding reproducibility I added the line

```
JMetalRandom.getInstance().setSeed(10L);
```

at the beginning of  the main-method.
So the main-method looks like this:

```
public static void main(String[] args) throws IOException {
      JMetalRandom.getInstance().setSeed(10L);

      PermutationProblem<PermutationSolution<Integer>> problem;
      Algorithm<List<PermutationSolution<Integer>>> algorithm;
      CrossoverOperator<PermutationSolution<Integer>> crossover;
      MutationOperator<PermutationSolution<Integer>> mutation;
      SelectionOperator<List<PermutationSolution<Integer>>, PermutationSolution<Integer>> selection;

      problem = new MultiobjectiveTSP("/tspInstances/kroA100.tsp", "/tspInstances/kroB100.tsp");
      [...]
```

Furthermore, I used the data (kroA100.tsp and kroB100.tsp) which was provided with the framework. I then ran the program once and my FUN.tsv started with the line

"93328.0 132453.0".

After I ran the program another time the FUN.tsv started with the line

"145839.0 101293.0".

However, I was wondering why this is happening. I thought that if I set the seed to the same value both times the results would be the same, since the random values which are generated during selection/mutation/... are the same (and this worked for me for non-permutation problems like the ZD1). Can you help me clarify my misunderstanding or my error?

Kind regards

**Comment by ajnebro** [comment]
You are right. Let me check why this happens.

**Comment by matthieu-vergne** [comment]
It would be nice to have a general test which check all the algorithms, and which passes only when reproducibility is confirmed everywhere (and fails by telling which algorithm is not reproducible).

**Comment by ajnebro** [comment]
The point is that most of algorithm components are based on random numbers. As the `JMetalRandom` class is a singleton, one way to reproduce results is to fix the seed. I use this extensively when debugging, but I don't know yet why fixing the seed does not work when running the `NSGAIITSPRunner` class. 

When testing classes involving randon numbers, I usually use mocking to set the values to be returned by the JMetalRandom` class.

**Comment by matthieu-vergne** [comment]
I never used mocking, so I am really unfamiliar with it... I should give it a try.

**Comment by hija** [comment]
@ajnebr: But if the JMetalRandom class is not the problem (at least as it seems to me right now), you cannot resolve the issue of not having reproducibility by mocking the JMetalRandom class, can you?

---
I had a look into some files and there were some files which did not use the JMetalRandom class, but I could not resolve the issue by using the JMetalRandom class in these classes... So I think that this even might be an internal issue like a functional call which uses a Random instance internally. Furthermore the error occurs also in the SMSEMOA class, so the reason might be in one of the superclasses...

**Comment by matthieu-vergne** [comment]
There is no reason for it to be solvable, unless there is another requirement to fulfill which is more important and fundamentally against reproducibility. Which I don't think there is. The only point is to be methodical, by analysing every pieces of the algorithm. And I don't see a better way to identify the piece at fault than to test the reproducibility of each of them. I only have a short time for internet for the moment, so I am a bit limited for my investigations right now, but I can try to look at it further when I find time.

**Comment by matthieu-vergne** [comment]
It appears that `MersenneTwisterGenerator` is used directly in `CoralReefsOptimization` (constructor + `generateCoordinates()`). There is also the simple `Random` which is used by:
- `DMOPSO`
- `CovarianceMatrixAdaptationEvolutionStrategy`
- `WFG`
- `Benchmark`

Besides that, I did not see any direct use, the others seem to pass through `JMetalRandom`. One should check out these direct uses to replace the invalid ones by proper uses of `JMetalRandom`.

**Comment by guv** [comment]
The singleton pattern is a very unfortunate choice for JMetalRandom since this prohibits parallel reproducible algorithm runs. Parallel algorithm runs will interfere unpredictably with the JMetalRandom singleton.
Hence, it is not possible to perform parallel evaluations for parameter tuning in the same JVM process.
A minimal invasive hotfix could be achieved using ThreadLocals [1] and have a "thread-local singleton".

[1] https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html

**Comment by ajnebro** [comment]
Hi @guv 

Let me check that.

**Comment by matthieu-vergne** [comment]
I agree with the point about the singleton. Was there a reason why to use the singleton? I see `JMetalRandom` like a mix between the need to make a generic interface and the will to make a wrapper (a single object but with the underlying implementation changing depending on the needs).

**Comment by ajnebro** [comment]
The typical reason of requiring a singleton: to ensure that there is a unique instance of the random number generator. This way we can reproduce the execution of an algorithm by fixing the seed.

What I understand from the post of @guv is that he wants to generalize the idea to a multi-thread context. However, currently we don't use explicitly threads in jMetal since we adopted the parallelStream() feature of Java 8, so I don't see how the seed could be set.

**Comment by guv** [comment]
Dropping th singleton and configuring the algorithm instance with a pseudo-random number generator would be the cleanest design. Though this is much more work than ThreadLocals fix mentioned above.

**Comment by ajnebro** [comment]
But if we drop the singleton, how do we ensure reproducibility when, for example, an operator creates its own instance of the random generator?

**Comment by guv** [comment]
The algorithm could pass its PRNG instance to its operators during initialisation.

**Comment by matthieu-vergne** [comment]
It should just not create it but request it. This is the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). The responsibility of the operator is to process its data, not to decide what is a relevant random strategy. You can give it a default instance, but enforcing this decision to be taken by the operator seems to me an arguable design.

The [singleton pattern](https://en.wikipedia.org/wiki/Singleton_pattern) should ensure that the same instance is used *everywhere* (in the JVM). This is for instance useful for an app which has many windows which should share all some common data. But in the case of running algorithms, there is no reason to think that two algorithm instances *must* share the same random generator. They could, but I don't see why it should be always true.

I am not sure the effort would be so huge to change it (an IDE helps a lot), but if the design seems better to be changed then it should be investigated. But first, the expected design should be confirmed. did not look the ThreadLocals either.

**Comment by ajnebro** [comment]
Although it could be possible to have more than one algorithm running at the same time, that's not the usual scenario and, in that case, it would be quite unusual to need to use different random generators. In my opinion, contemplating this possibility doesn't justify changing the current implementation.

**Comment by guv** [comment]
Just to be clear: my criticism of the singleton approach came from a real use case - parameter tuning of an NSGA-II algorithm with parallel evaluations on several different computation servers. Having to set up one JVM process per thread is very inconvenient and not memory consumption friendly.

Since good parameter configurations of EA or metaheuristics are largely problem-dependent and their choice is not immediately apparent, parameter tuning is an essential step to develop a working solver for a problem.

**Comment by ajnebro** [comment]
> Just to be clear: my criticism of the singleton approach came from a real use case - parameter tuning of an NSGA-II algorithm with parallel evaluations on several different computation servers. Having to set up one JVM process per thread on each server is very inconvenient and not memory consumption friendly.

I'm not sure to get the point. Currently, we can make the evaluations of NSGA-II (and other algorithms) in parallel in multicore computers using the multi-threaded evaluator (only a JVM is needed) and in clusters using an [Spark-based evaluator](http://link.springer.com/chapter/10.1007%2F978-3-319-54157-0_2) (a JVM will be need per node). In the second clase is clear that JVM will have its copy of the jMetalRandom object. 

Anyway, assuming that you could have a local copy of the random generator per thread, how would you manage aspects such as thread scheduling in order to reproduce an execution?

> Since good parameter configurations of EA or metaheuristics are largely problem-dependent and their choice is not immediately apparent, parameter tuning is an essential step to develop a working solver for a problem.

I agree. This is an open issue, and some people are using tools (e.g. irace, ParamILS) for that purpose.

**Comment by guv** [comment]
The point in one sentence: I need reproducible algorithm runs using specified seeds such that each run has exactly the same sequence of random numbers.

This fails with a global PRNG singleton that is shared between all parallel executing algorithm runs in the same JVM. It would be guaranteed if each algorithm had its own PRNG instance that can be seeded in the algorithm initialisation/setup.

**Comment by ajnebro** [comment]
If I understand well, you are executing different instances of NSGA-II in a same JVM (I undestood that you had a single algorithm with the evaluations in parallel). If so, I see now your problem.

**Comment by guv** [comment]
Correct, several algorithm runs in the same JVM in parallel.
In the tuning setup the evaluations are done sequentially.

**Comment by ajnebro** [comment]
And you want all the NSGA-IIs have the same seed for the random generator.

**Comment by guv** [comment]
Yes, each algorithm run (e.g. NSGA-II) must have the same sequence of random numbers.

**Comment by matthieu-vergne** [comment]
> Although it could be possible [...] that's not the usual scenario

I remember someone said that this kind of reasons shows how limited is the global picture of the one who uses them. Don't remember who...

So should we discuss a revision of the `JMetalRandom` class?

**Comment by ajnebro** [comment]
I don't remember who said that :P, and I disagree about that assessment :)

Suggestions are always welcome.

**Comment by matthieu-vergne** [comment]
Well, at least to solve this problem you just need to make the constructor public.

**Comment by ajnebro** [comment]
But ... if we make the constructor public, it doesn't make sense to use a Singleton.

**Comment by matthieu-vergne** [comment]
I agree. But for retrocompatibility reason, you can make it deprecated and take your time to update the various uses of the `getInstance()` static method. Concretely, you don't need more than that to make it works. Then, the point is to think further about how to revise the ones which use the deprecated method, one by one.

**Comment by ajnebro** [comment]
Hmmm, that should be easy (a simple refactoring), but I'm not sure if doing that change will solve the issue of @guv because, for example, when an operator needs to use the random generator, what instance will it take?

**Comment by matthieu-vergne** [comment]
No, it only solves the problem of the inability to use several random generators in parallel in the same JVM, of course. The point is that we don't need to change everything, we can first revise the few operators and algorithms that @guv needs to be refined. It is not about a significant design change. I did not check the use of the method yet (my personal computer reboot randomly when I am a bit too intensive... like using Firefox for too long... so I cannot do much now) but simple calls to `getInstance()` can be replaced by the use of a variable instead, variable initiated as any other parameter of the algorithm/operator/whatever through a setter or constructor argument. Normally, it should not be a hard thing to change.

**Comment by matthieu-vergne** [comment]
Maybe @guv  could list the few stuff he uses?

**Comment by matthieu-vergne** [comment]
I plan to look at it tomorrow on my professional computer, so I can come up with more concrete comments.

**Comment by ajnebro** [comment]
After thinking about this issue, this is what I would do:
* Clone the jMetal repository
* Implement a new jMetalRandom class without the singleton pattern (i.e. a plain Java class)
* Modify NSGA-II to receive instances of this class (by a constructor parameter or a setter)
* Write a version of the selection, mutation and crossover operators to receive the random generator of the algorithm that will use them.
* Repeat the last step in all the classes using the random generator. For example, `DefaultDoubleSolution` is NSGA-II is being used to solve continuous problems.

As I said before, I don't think that jMetal has limitations because it doesn't allow to deal with @guv's issue. In particular cases like this, just take the source code and modify it according to your needs.

**Comment by ajnebro** [comment]
Of course, if we find a solution that is more general than the current implementation and that solution can cope with this issue, no problem in adopting it.

**Comment by matthieu-vergne** [comment]
> As I said before, I don't think that jMetal has limitations because it
doesn't allow to deal with @guv <https://github.com/guv>'s issue.

The extreme cases are the ones which show the limitations of a design.
Still, extreme cases are -although not frequent- still relevant cases. If
you think this is not your design which is limited, then it is @guv who
uses a class that he should not. It is not my opinion (at all) and your
vision makes me feel really frustrated, but you are the one who implements
it.

> Of course, if we find a solution that is more general than the current
implementation and that solution can cope with this issue, no problem in
adopting it.

Somehow you contradict what you said just before: if the case should not be
covered, it is because the restriction is made on purpose, so accepting a
more general implementation which covers it means that actually this
restriction is arbitrary. Surely our disagreement comes again from our
different visions, so I won't dig again in that here and be more concrete.

As I said, to unlock the ability to reuse `JMetalRandom`, you only need
this:
- make the constructor of `JMetalRandom` public
- make its `getInstance()` deprecated

With an IDE, you can spot and go immediately to the various uses of (now
deprecated) `getInstance()`, such that changing algorithms or operators
using it is long but really easy and can be done progressively. For each of
them, simply:
- replace the deprecated calls to `getInstance()` by the use of a
`randomGenerator` field (or variable if only used in the constructor)
- add a `randomGenerator` parameter to the constructor (usually one)
- initialise the field/variable with the parameter
- add a deprecated constructor without this parameter which calls the
previous constructor by using `JMetalRandom.getInstance()` as a default

This is the last line which allows it to be progressive, because it keeps
the old code but spread the relevant warnings for future refinements. In
particular, because the last constructor is deprecated too, its use of the
deprecated `getInstance()` is not considered as a mistake, so no warning
appears for this one. Good thing because it is not supposed to be another
way. Moreover, because these changes make the old constructor (i.e. without
`randomGenerator`) deprecated, it adds further warnings for the classes
which instantiate or extend the refined class, which allow again to easily
spot the required refinements. So it can be done progressively until
everything is refined. The only thing to pay attention to, is to refine
first the "leafs", so operators and algorithms which do not use themselves
instances building on the deprecated `getInstance()`. There is a way to
spot them easily, but it add quite a bit of work. Still easy, but long
again.

What makes it hard are actually the builder mania (*) of JMetal, because
builders need to add a new method rather than change an existing one, so it
is harder to spot the relevant places to change. Moreover, because it
relates to simple methods (not constructors), a deeper analysis of how the
builder is used can be required to spot the right place to use the new
method. And because there is builders everywhere, thre is difficulties
everywhere.The good thing is that we can still do the easy stuff above,
because it only leads the builders to use deprecated constructors.

(*) I call that "builder mania" because you have useless builders: some are
unused, like the `EnvironmentalSelection.Builder`, and because your
algorithms usually need all their parameters they only need few
constructors, making a builder irrelevant. Indeed, builders help to deal
with the multiplication of constructor variants. What you seem to need is
often method cascading, which relates to fluent interfaces, not builders.
They can be mixed because builders often use cascading methods, but they do
not require them, so making builders to have cascading method is a mistake.
I use method cascading in my version of the experiments, which is not
finalised yet but I think you will like it, but no builder there.

**Comment by ajnebro** [comment]
I agree with the builder mania. Sometimes I have though in making all of them optional. 

>The extreme cases are the ones which show the limitations of a design.
Still, extreme cases are -although not frequent- still relevant cases. If
you think this is not your design which is limited, then it is @guv who
uses a class that he should not. It is not my opinion (at all) and your
vision makes me feel really frustrated, but you are the one who implements
it.
>> Of course, if we find a solution that is more general than the current
implementation and that solution can cope with this issue, no problem in
adopting it.
>Somehow you contradict what you said just before

I'm affraid that I'm not explaining me well. What I say is that if we find a satisfactory way of to cope with this issue so that everything is covered, then let's adopt it but, if not, as it seems, I would modify a personal copy the project on my own. I don't see that the change you propose be a solution to the issue, because you have to do all the steps I mentioned before.

And of course the jMetal has limitations; I could write a very long list ... For example, all the objectives are intended to be minimized, so you have to transform the maximization objectives into minimization ones, so to obtain the right objective values the result files must be processed offine. It's not difficult to do, but to some people this is cumbersome. In a project we adopted an ad-hoc solution for this, but I decided not to incorporate it in the master branch because it would require too much changes.  

Someone said (A.S. Tanenbaum, if I remember well) that it is better a good simple design which covers the 95% of the cases than a perfect complex design covering everything. I always liked this idea.

**Comment by matthieu-vergne** [comment]
> I don't see that the change you propose be a solution to the issue,
because you have to do all the steps I mentioned before.

I don't really see your point. If you say that a more generic
implementation which can cover the case of @guv is OK, I propose you
exactly that. From my previous post, remove all the deprecated annotations
and all what it is about is making them more generic by adding the
possibility to choose a custom random generator. Legacy code still works,
but for the ones you change, you add the possibility to specify your own
random generator. It is pure extension of functionality.

Let's go this way: I will propose pull requests which make operators more
generic, without touching algorithms using them nor JMetalRandom. If you
accept them, then after that I can propose the same thing for algorithms.
Without touching their builders. And so on. So you can actually see that it
is really progressive and does not break legacy code, only make these
classes more generic.

Is that OK for you?

**Comment by matthieu-vergne** [comment]
By the way, I also agree with that:

> Someone said (A.S. Tanenbaum, if I remember well) that it is better a
good simple design which covers the 95% of the cases than a perfect complex
design covering everything.

The critics is not on doing everything, it is on complexity. The whole
point is that a more generic design is less complex because uses less
assumptions. One should not confuse between a simple class which does the
necessary thing and a complex class which does more than actually required.
I target the first one, not the second.

**Comment by ajnebro** [comment]
> Let's go this way: I will propose pull requests which make operators more generic, without touching algorithms using them nor JMetalRandom. If you accept them, then after that I can propose the same thing for algorithms. Without touching their builders. And so on. So you can actually see that it is really progressive and does not break legacy code, only make these classes more generic. Is that OK for you?

You know the anwser ;)

**Comment by ajnebro** [comment]
By the way, I didn't mentioned that I'm not a big fan of the Singleton patttern, but until know it has worked well enough in the case of the jMetalRandom class.

**Comment by matthieu-vergne** [comment]
So here is the first pull request: #198
